File: apps/api/src/integrations/_mocks/dropboxsign.ts
import type { HttpRequest, HttpResponse } from '../_http/client.js';
import { jsonResponse, consumeFailureFor } from './transport.js';
import { getMockStore } from './store.js';

// Mock Dropbox Sign endpoints:
//   POST /v3/signature_request/remind/{id}     — send reminder
//   POST /v3/signature_request/send_with_template — create from template
//   POST /v3/signature_request/cancel/{id}     — void

export async function handleDropboxsign(req: HttpRequest, path: string): Promise<HttpResponse> {
  const fail = consumeFailureFor(`dropboxsign:${req.method} ${path}`);
  if (fail) return fail;

  const store = getMockStore();

  const remindMatch = path.match(/^\/v3\/signature_request\/remind\/([^/]+)$/);
  if (req.method === 'POST' && remindMatch) {
    const id = remindMatch[1] ?? '';
    if (!store.contracts.has(id)) return jsonResponse(404, { error: 'signature_request not found' });
    store.contract_reminders.push({ signature_request_id: id, at: new Date().toISOString() });
    return jsonResponse(200, { signature_request_id: id, reminded: true });
  }

  if (req.method === 'POST' && path === '/v3/signature_request/send_with_template') {
    const body = req.body as { template_id: string; recipient: string; title: string };
    const id = `sig_${Date.now()}`;
    store.contracts.set(id, {
      id,
      title: body.title,
      recipient: body.recipient,
      status: 'sent',
      sent_at: new Date().toISOString(),
      signed_at: null,
    });
    return jsonResponse(201, { signature_request_id: id, status: 'sent' });
  }

  const cancelMatch = path.match(/^\/v3\/signature_request\/cancel\/([^/]+)$/);
  if (req.method === 'POST' && cancelMatch) {
    const id = cancelMatch[1] ?? '';
    const c = store.contracts.get(id);
    if (!c) return jsonResponse(404, { error: 'not found' });
    c.status = 'expired';
    return jsonResponse(200, { signature_request_id: id, cancelled: true });
  }

  return jsonResponse(404, { error: `dropboxsign: unmocked ${req.method} ${path}` });
}

File: apps/api_py/src/linkbook/integrations/mocks/dropboxsign.py
"""Dropbox Sign mock — mirrors apps/api/src/integrations/_mocks/dropboxsign.ts."""

from __future__ import annotations

import re
import time
from datetime import datetime, timezone

from ..http import HttpRequest, HttpResponse
from .store import MockContract, get_mock_store
from .transport import consume_failure_for, json_response


async def handle(req: HttpRequest, path: str) -> HttpResponse:
    fail = consume_failure_for(f"dropboxsign:{req.method} {path}")
    if fail is not None:
        return fail
    store = get_mock_store()

    m = re.match(r"^/v3/signature_request/remind/([^/]+)$", path)
    if req.method == "POST" and m:
        sig_id = m.group(1)
        if sig_id not in store.contracts:
            return json_response(404, {"error": "signature_request not found"})
        store.contract_reminders.append(
            {"signature_request_id": sig_id, "at": datetime.now(timezone.utc).isoformat()}
        )
        return json_response(200, {"signature_request_id": sig_id, "reminded": True})

    if req.method == "POST" and path == "/v3/signature_request/send_with_template":
        body = req.body or {}
        new_id = f"sig_{int(time.time() * 1000)}"
        store.contracts[new_id] = MockContract(
            id=new_id,
            title=body.get("title", ""),
            recipient=body.get("recipient", ""),
            status="sent",
            sent_at=datetime.now(timezone.utc).isoformat(),
            signed_at=None,
        )
        return json_response(201, {"signature_request_id": new_id, "status": "sent"})

    m = re.match(r"^/v3/signature_request/cancel/([^/]+)$", path)
    if req.method == "POST" and m:
        sig_id = m.group(1)
        c = store.contracts.get(sig_id)
        if c is None:
            return json_response(404, {"error": "not found"})
        c.status = "expired"
        return json_response(200, {"signature_request_id": sig_id, "cancelled": True})

    return json_response(404, {"error": f"dropboxsign: unmocked {req.method} {path}"})
